Agg doc (#713)
partiql/partiql-lang-kotlin: docs/user/AggFunc.md

# Aggregate Function

> Aggregate function outputs a single result value from a collection of input values.

Reviewer: ```suggestion
Aggregate functions output a single result value from a collection of input values.
```

Unlike SQL, `GROUP BY` is not required to use the aggregate function. 

> Unless otherwise specified, aggregate function will skip `NULL` and `MISSING`

Reviewer: ```suggestion
Unless otherwise specified, aggregate functions will skip `NULL` and `MISSING`
```

## AVG

Returns the average (arithmetic mean) of the input expression values.

Signature
: `AVG: COLLECTION OF NUMERIC TYPE -> DECIMAL OR NULL`

Header
: `AVG(expr)`

Purpose

> : Given a collection of numeric type values, evaluates the arithmetic mean of the input expression values. 

Reviewer 2: I think this can be consolidated to just **Description** and **Signature**. The description and purpose are the same, and the HEADER doesn't tell us any more than the signature. Signature should be plural.

Consider,

## AVG

### Description
Given a collection of numeric type values, `AVG` evaluates the arithmetic mean of the input expression values.

### Signatures
`AVG(collection<numeric>) -> decimal or NULL`

Author: Addressed in the new commit.

Example
: 

```SQL
    AVG(<< 1,2,3 >>) -- 2 (Decimal type with Precision of 1 and Scale of 0)
    AVG(<< 1,2,3,'1'>>) --error
    AVG(<<MISSING>>) -- NULL

>     AVG(<< 1,2,3, MISSING >>) -- 2

Author: Ideally yes, but I hesitate to designate this doc to partiql-docs at this point because I felt what we have implemented in the Lang-Kotlin may not be exactly what we desire for spec. 

For example, one may argue that `AVG(<< 1, 2, 3 >>` should return 2 as an integer instead of `decimal(1,0)`. 

My original intention for this document is to keep track of the current Lang-Kotlin behavior so that 1) customers has a place to refer to 2) we can cross reference the current implementation with sql spec and other popular sql implementations.

>     AVG(<< 1,2,3, MISSING >>) -- 2

Author: Maybe the drafts folder under partiql-docs?

>     AVG(<< 1,2,3, MISSING >>) -- 2

Reviewer 2: Understood. It seems then that the aggregate functions need to be formally defined along with their conformance to the SQL specification

```

## COUNT

Returns the number of “rows” in the input.

Signature
: `COUNT: STAR -> INT8`
: `COUNT: Any DATATYPE -> INT8`

Header
: `COUNT(*)`
: `COUNT(expr)`

Purpose
: Given a collection of variable in any date type, calculate the number of value in the collection.
Note: `COUNT(expr)` will skip the null and missing value, but COUNT(*) will not. 

Example
: 

```sql
COUNT(<< 1,2,3 >>) -- 3
COUNT(<< 1,2,3,<<'1', {'a':1} >> >>) -- 4
-- COUNT(expr) ignores null and missing
SELECT COUNT(a) FROM <<MISSING>> AS a -- 1 
SELECT COUNT(a) FROM <<MISSING, 1, 2>> AS a -- 2 

-- COUNT(*) includes null and missing
SELECT COUNT(*) FROM <<MISSING>> AS a -- 1 
SELECT COUNT(*) FROM <<MISSING, 1, 2>> AS a -- 3 
```

## MIN/MAX

> Returns the MIN/MAX value of the input expression values.

Reviewer 2: Include information regarding the comparator. Is the comparator consistent with `ORDER BY`?

Author: Addressed in the new commit. The comparator is consistent with `ORDER BY` with one exception, it skips `NULL` and `MISSING`.

Signature
: `MIN: Any DataType -> DATATYPE`
: `MAX: Any DataType -> DATATYPE`

Header
: `MIN(expression)/MAX(expression)`

Purpose

> : Given a collection of variable in any date type, find the minimum/maximum value in the collection.

Reviewer: ```suggestion
: Given a collection of variables in any date type, find the minimum/maximum value in the collection.
```

: In general: Boolean < Number < Text < Blob/Clob < List < Struct < Bag

Example
: 

```SQL

>     MIN(<<1,2,3>>) -- 1
>     MIN(<<'1',2,3>>) -- 2
>     MIN(<<'1',2,3, MISSING>>) -- 2
>         
>     MAX(<<1,2,3>>) -- 3
>     MAX(<<'1',2,3>>) -- '1'
>     MAX(<<'1',2,3, MISSING>>) -- '1'

Reviewer 2: ```suggestion
MIN(<< 1, 2, 3 >>)            -- 1
MIN(<< '1', 2, 3 >>)          -- 2
MIN(<< '1', 2, 3, MISSING >>) -- 2

MAX(<< 1, 2, 3 >>)            -- 3
MAX(<< '1', 2, 3 >>)          -- '1'
MAX(<< '1', 2, 3, MISSING >>) -- '1'
```

```

## SUM

Returns the sum of the input expression values.

Signature
: `SUM: COLLECTION OF NUMERIC TYPE -> DECIMAL OR NULL`

Header
: `SUM(expr)`

Purpose
: Given a collection of numeric type values, evaluates the sum of the input expression values.

Example
: 

```SQL

>     SUM(<< 1,2,3 >>) -- 6
>     SUM(<<1,2,`3.0d0`>>) -- 6.0
>     SUM(<<1,2,`3.0e0`>>) --6.0e0
>     SUM(<< 1,2,3,'1'>>) --error
>     SUM(<<MISSING>>) -- NULL

Reviewer 2: ```suggestion
SUM(<< 1, 2, 3 >>)         -- 6
SUM(<< 1, 2,`3.0d0` >>)    -- 6.0
SUM(<< 1, 2, `3.0e0` >>)   -- 6.0e0
SUM(<< 1, 2, 3, '1' >>)    -- !! ERROR !!
SUM(<< MISSING >>)         -- NULL
```

# Aggregate Functions

Aggregate functions output a single result value from a collection of input values. 

Unless otherwise specified, aggregate functions will skip `NULL` and `MISSING`.

## AVG

Description
: Given a collection of numeric type values, evaluates the arithmetic mean of the input expression values.

Signatures
: `AVG(collection<NUMERIC | NULL | MISSING>) -> DECIMAL or NULL`

Example
: 

```SQL

> AVG(<< 1, 2, 3 >>)          -- 2 (`decimal(1, 0)`

Reviewer 2: ```suggestion
AVG(<< 1, 2, 3 >>)          -- 2 `decimal(1, 0)`
```
